### DISCUSSION TITLE
Mutiny: How to cancel a step in a chained pipeline?

### DISCUSSION THREAD

**Discussion by flowt-au** [quarkusio/quarkus#28394] [Q&A]
Hi.
I have a question about cancelling in a Uni pipeline. I have a sequence where I do something before the main thing, do the main thing, do something after the main thing. The "before" step needs to be cancellable so my business logic can say "No, you cant do the main thing for some reason". 

If I have a Uni pipeline like this pseudocode:
```java
this.beforeSomething(arg1)
    .onCancellation(handleBeforeCancellation())
    .chain(resultFromBefore -> doSomething(resultFromBefore))
    .onCancellation(handleCancellation())
    .chain(resultFromSomething -> afterSomething(resultFromSomething))
    .onCancellation(handleAfterCancellation())
    .subscribe().with(
      res -> System.out.println("Test flow success=" + res),
      failure-> System.out.println("Test flow failure=" + failure)
    );
```
Each of my methods returns a `Uni<String>`

how do I emit a "cancel" from my methods so the onCancellation will fire? Something like:
```java
Uni<String> beforeSomething (String data) {
  if (allOk) {
    return Uni.createFrom().item("DataForSomething");
  } else {
    // A method like this fake cancelItem()...
    return Uni.createFrom().cancelItem("DataFromCancellation");
  }
}
```

I couldn't find anything in the docs.

Thanks,
Murray

[a comment by an automated account is not shown]

**Reply by jponge** [reply]
Subscriptions are forwarded by operators from the source to the downstream subscriber. If you have an operator in the middle that triggers a `subscription.cancel()`, then a cancellation signal is sent up to the source publisher so it eventually stops emitting events.

You might want to look at [onSubscription()](https://javadoc.io/static/io.smallrye.reactive/mutiny/1.7.0/io/smallrye/mutiny/Uni.html#onSubscription()) to catch the subscription somewhere in your pipeline, and pass it somehow to your code so it can perform a cancellation request.

**Reply by flowt-au** [reply]
Thanks for your reply @jponge 
I realised I was misunderstanding the role that a cancellation plays.
I found a way to do this with just item and failure which I will post once I am happy it is a solution.
Cheers,
Murray

**Reply by flowt-au** [reply]
@jponge As mentioned, I have tested my approach and documented it here for comment and feedback.
[https://github.com/quarkusio/quarkus/discussions/28621](https://github.com/quarkusio/quarkus/discussions/28621)
Thanks again, Julien.
